libnmstate/metadata.py

```diff
@@ -181,3 +193,73 @@ def _generate_route_metadata(desired_state):
                 iface_state[Interface.IPV6][ROUTES].append(route)
             else:
                 iface_state[Interface.IPV4][ROUTES].append(route)
+
+
+def _generate_dns_metadata(desired_state, current_state):
+    """
+    Save DNS configuration to choose interface as metadata.
+    """
+    # Only support up to 2 name servers.
+    servers = desired_state.config_dns.get(DNS.SERVER, [])[:2]
+    searches = desired_state.config_dns.get(DNS.SEARCH, [])
+    # Some interfaces might old DNS configuration which is not included
+    # in desired_state, we need to add them into desired_state in order to
+    # clear its DNS configurations
+    client = nmclient.client()
+    old_dns_ifaces = nm_dns.get_dns_config_iface_names(
+            nm_ipv4.acs_and_ip_profiles(client),
+            nm_ipv6.acs_and_ip_profiles(client)
+    )
+    _include_name_only_iface_state(
+        desired_state, current_state, old_dns_ifaces)
+    if not servers and not searches:
+        return
+    # The route need to be merged before checking default gateways
+    desired_state.merge_route_config(current_state)
```

Comment:
It is already done in `netapplier` before reaching the metadata processing. I do not think it is needed here.